### PR TITLE
Add PHPBrew to VVV

### PR DESCRIPTION
**Pull request by ndisidore**
## Why
This is my initial attempt at adding PHPBrew to VVV. Over the course of development, there were times where I more closely wanted to mimic our production environment, which may not be the latest version of PHP, may or may not have memcached installed, etc. Rather than just install a single version of PHP, I stumbled upon PHPBrew and thought it would be beneficial to integrate into VVV. PHPBrew allows you to easily build php versions on the fly, and turn on/off extensions like memcached, imagemagic, and APC.

More information on PHPBrew, along with it's command list, can be found here: https://github.com/phpbrew/phpbrew

## How to use
This installs phpbrew to /opt/phpbrew and registers a new binary, `phpswap`, that is a wrapper for phpbrew functionality. Switching PHP versions is as simple as issuing a single command from terminal
`phpswap 5.3.28`
This looks for the version installed locally, and, it doesn't find it, grabs the source and builds from scratch with the set list of extensions.
Restart the fpm process for the currently running version of PHP with `phpswap fpm-restart`
You can switch the version persist across reboots by using the -s flag, e.g. `phpswap -s 5.3.28`
Switch back to the system version with `phpswap system`.

## Notes
* Even though the script executes `phpbrew use/switch $PHP_VER`, it doesn't seem to register unless you run the command separately, e.g. `phpswap 5.5.20 && phpbrew use 5.5.20` or `phpswap -s 5.5.20 && phpbrew switch 5.5.20`
* After enabling/disabling extensions, issue `phpswap fpm-restart` for the changes to take effect on the web server

### CODE REVIEW

**Comment by joelworsham** [comment]
+1

**Comment by jeremyfelt** [comment]
**Finally** coming back around to this. Thank you @ndisidore for all this work! Sorry to keep you hanging so long. I got a chance to try it out tonight and it's pretty amazing.

I'm currently exploring the steps required to get to the the beginning and the end of the WordPress range:

* While `phpbrew` supports the `--old` flag to install versions earlier than 5.3, PHP has moved the release files for 5.2.4 and other old versions to `museum.php.net`, which is not mirrored. This results in a quick failure. I setup my own mirror at `vvv.dev/distributions/` and specified it with `phpbrew install --old --mirror http://vvv.dev 5.2.4`. This started to work, but `make` failed when building openssl. I'm guessing there is still a workaround, which would be nice to find for core development.
* Switching to a mid-range version was easy. `phpswap 5.3.20` was seamless. I didn't even have to do anything extra to restart the service.
* `phpswap` doesn't support installing (I think) the unstable 7 releases. A few related notes here in my attempt to get it going:
    * I installed with `phpbrew install next as php-7.0.0` so that the package was available and then tried `phpswap 7.0.0`. I got an error related to php-fpm, which apparently was not installed by default.
    * I then tried `phpbrew install next as php-7.0.0 +fpm` to rebuild with the fpm variant. This got me closer, but there's some kind of config issue due to a missing pool. It's here that I'm done for the night, so I'm guessing it's just an adjustment somewhere.

I think we're close, and I'm looking forward to having a place to test 7 out as well. I also think the comments on #586 about quickly testing reported errors in other versions are important. This goes along with client work and with core development.

I'm going to work toward getting this merged into an updated version of the feature branch very soon. I'd like to see this as we complete the feature:

1. Standard PHP 5.3.x -> 5.6.x support via the phpswap command inside the VM.
2. Support for PHP 7 via the phpswap command inside the VM.

And longer term:
1. Documented steps to get 5.2.4 working with phpbrew.
2. Support for PHP 5.2.4 via the phpswap command inside the VM. I'm honestly not worried about any of the other 5.2.x releases, only 5.2.4 as it will be very useful for a small amount of testing.

**Comment by helgatheviking** [comment]
:100: Think I just got burned again by PHP errors that I didn't anticipate with VVV. Would love to see this get merged in.

File: provision/provision.sh [hunk 1]
```diff
@@ -178,12 +199,30 @@ if [[ $ping_result == "Connected" ]]; then
...
+		PHPBREW_EXP="export PHPBREW_ROOT=$PHPBREW_ROOT"
+		PHPBREW_HME="export PHPBREW_HOME=$PHPBREW_HOME"
+		if [ ! -f $PHPBREW_ROOT/init ]; then touch $PHPBREW_ROOT/init; fi
+		grep -q "$PHPBREW_EXP" $PHPBREW_ROOT/init || echo "$PHPBREW_EXP" >> $PHPBREW_ROOT/init
+		grep -q "$PHPBREW_HME" $PHPBREW_ROOT/init || echo "$PHPBREW_HME" >> $PHPBREW_ROOT/init
+		PHPBREW_BASHRC="source $PHPBREW_ROOT/bashrc"
+		grep -q "$PHPBREW_BASHRC" ~/.bashrc || echo "$PHPBREW_BASHRC" >> ~/.bashrc
+		echo "$PHPBREW_EXP
+$PHPBREW_HME
+source $PHPBREW_ROOT/bashrc" > /etc/profile.d/phpbrew.sh
```
Comment (cfoellmann) [inline provision/provision.sh]: formatting...

File: config/bash_profile [hunk 2]
```diff
@@ -37,4 +37,9 @@ eval "$(grunt --completion=bash)"
...
 
 # PHPCS path
-export PATH="$PATH:/srv/www/phpcs/scripts/"
\ No newline at end of file
+export PATH="$PATH:/srv/www/phpcs/scripts/"
+
+# PHPBrew Exports
+export PHPBREW_ROOT=/opt/phpbrew
+export PHPBREW_HOME=/opt/phpbrew
+. /opt/phpbrew/bashrc
```
Comment (cfoellmann) [inline config/bash_profile]: lets add a "new line" at the end when we are on it.

File: config/homebin/phpswap [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+# Restart phpbrew's fpm
+echo "Restarting fpm so the changes take effect"
+brew_fpm_off
+BREW_FPM_START=$(brew_fpm_on $PHP_VER | tee /dev/tty)
+
+if ! [[ $BREW_FPM_START == *"failed"* ]]; then
+	echo "$(tput setaf 2)PHP version swapped to $PHP_VER$(tput sgr 0)"
+else
+	echo "$(tput setaf 1)ERROR: There was a problem switching PHP versions$(tput sgr 0)"
+fi
```
Comment (cfoellmann) [inline config/homebin/phpswap]: lets add a "new line" at the end

**Comment by cfoellmann** [comment]
@ndisidore change the destination branch to `develop`, please. (I know that is only possible by creating a new PR). Also make sure you merge in the current changes from `develop`
THX

@jeremyfelt is this ok? 
We (I) need to merge `master` -> `develop`

**Comment by cfoellmann** [comment]
@ndisidore send all new changes to the `develop` branch, please. I will keep the `feature/phpbrew` branch current to give people a basis to test.

**Comment by eteubert** [comment]
Any updates on this end? I just tried switching to the feature/phpbrew branch and switching to PHP 7 using `phpswap` but could not get it to work. Here's the log: https://gist.github.com/eteubert/f1e68dc04b8b510673fd

**Comment by ihorvorotnov** [comment]
+1 What's the status? PHP 7 is already out and spreading across the webs. Would be nice to have it in VVV too.

**Comment by aaroncampbell** [comment]
I don't know if phpbrew is the way to go or not, but I'd LOVE to see PHP 7 be a simple option (or even the default) one way or another.

**Comment by GaryJones** [comment]
I've used phpbrew on the local machine, and it's really handy. Great to see it as part of VVV too. When do we get to have it, @jeremyfelt? :-)

File: config/homebin/phpswap [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+	# Switch the PHP version
+	phpbrew $SWAP_TYPE $PHP_VER
+	# Install extensions
+	phpbrew ext install memcache
+	phpbrew ext install imagick
+
+	# Xdebug is special because the 2.3 branch drops support for PHP 5.2/5.3
+	[[ $PHP_VER == "5.2"* || $PHP_VER == "5.3"* ]] && { XDEBUG_VER=2.2.7; true; } || XDEBUG_VER=stable
+	phpbrew ext install xdebug $XDEBUG_VER
```
Comment (GaryJones) [inline config/homebin/phpswap]: For PHP 7, XDebug 2.4+ is needed, the latest of which is 2.4.0RC4.

File: config/homebin/phpswap [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+	fi
+
+	# Make sure Ubuntu's php-fpm is stopped
+	sudo service php5-fpm stop
+
+	# Switch the PHP version
+	phpbrew $SWAP_TYPE $PHP_VER
+	# Install extensions
+	phpbrew ext install memcache
+	phpbrew ext install imagick
```
Comment (GaryJones) [inline config/homebin/phpswap]: For PHP 7, right now, imagick stable fails to install.  imagick 3.4.0RC6 seems to install.

File: config/homebin/phpswap (hunk 4 above)
Comment (GaryJones) [inline config/homebin/phpswap]: Xdebug 2.4.0 has now been released

File: config/homebin/phpswap (hunk 5 above)
Comment (GaryJones) [inline config/homebin/phpswap]: [Imagick](http://pecl.php.net/package/imagick) 3.4.0 has been released today.

**Comment by mishtershmart** [comment]
Howdy,
Would someone mind giving a quick explanation as to how I get this happening? Do I need to check out this PR, or should this already be available in the most up-to-date version of VVV? 

Thanks!

**Pull request merged**